Handle SIGINT

tresf opened the pull request:
Small PR to allow Ctrl + C to close LMMS.

messmerd commented:
Looks like calling `qApp->exit(3);` isn't allowed from the signal handler. See: https://doc.qt.io/qt-5/unix-signals.html

tresf commented:
> Looks like calling `qApp->exit(3);` isn't allowed from the signal handler. See: https://doc.qt.io/qt-5/unix-signals.html

They say that, but [`exit()` is explicitly thread safe](https://doc.qt.io/qt-6/qcoreapplication.html#exit), so I think that's out of caution.

It's a good idea anyhow incase we need to run some custom code.  PR updated with signal/slots.

tresf commented on include/GuiApplication.h:
```diff
@@ -67,11 +67,17 @@ class LMMS_EXPORT GuiApplication : public QObject
 	AutomationEditorWindow* automationEditor() { return m_automationEditor; }
 	ControllerRackView* getControllerRackView() { return m_controllerRackView; }
 
+	void sendInterrupt(int signal);
```
There's a slight bit of redundancy by calling it "sendInterrupt()` to then double-check that the code was `SIGINT`.  I'm fine genericizing and simplifying this further if needed.  I'd rather call this `sendSignal`, but in Qt-land, the word "signal" has its own meaning and I was for a loss of words.

tresf replied:
Resolving, since this code needs to change before merge.

tresf commented on src/core/main.cpp:
```diff
@@ -99,6 +99,16 @@ void signalHandler( int signum ) {
+		exit(3);
+	}
+}
+
```
I considered removing this and binding `signal(SIGINT` directly to the GuiApplication function, but I wanted to also capture the interrupt from CLI and I thought this was the cleanest way to handle both.

tresf commented on src/gui/GuiApplication.cpp:
```diff
@@ -240,6 +244,19 @@ void GuiApplication::childDestroyed(QObject *obj)
+	switch(signal) {
+		case SIGINT:
+			// add last minute resource clean-up goes here
+			qApp->exit(3);
```
`exit()` seems abrupt, but if we call `quit()`, `MainApplication` will start asking questions which no one wants.  I checked the clean-up code and it primarily deletes a recovery file, which we can't possibly know if should or should not be done, so I believe this will run any necessary destructors from a memory perspective and be as clean as possible.  If this changes down the road, I would advise we move this slot into `MainApplication`, but I kept it here for simplicity.

tresf commented on src/core/main.cpp:
```diff
@@ -99,6 +99,16 @@ void signalHandler( int signum ) {
 }
 #endif
 
+void interruptHandler(int code) {
+	using namespace lmms::gui;
+
+	if(getGUI() != nullptr) {
+		getGUI()->sendInterrupt(code);
+	} else {
+		exit(3);
+	}
+}
+
```
I considered removing this and binding `signal(SIGINT` directly to the GuiApplication function, but I wanted to also capture the interrupt from CLI and I thought this was the cleanest way to handle both.

tresf replied:
I'm no longer catching the SIGINT in this fashion, which does leave the terminal abandoned.  Since unix signal handlers are so damn strict, quoting:

> It's not safe to call a non-async safe function from within the signal handler

... then we can't even check to see if we have initialized `lmms::gui::getGUI()`, we can only write to an arbitrary file handle and hope it's received on the other end.    Resolving this for now, but I wouldn't be surprised if this PR is failed for this reason alone.

sakertooth commented:
I think we should use `QSocketNotifier` and `sigaction`/`socketpair` as mentioned in the Qt guide by @messmerd. It feels like that's what the proper way to handle UNIX signals in this scenario.

From the Qt guide:
> One simple way to make this happen is to declare a socket pair in your class for each Unix signal you want to handle. The socket pairs are declared as static data members. You also create a [QSocketNotifier](https://doc.qt.io/qt-6/qsocketnotifier.html) to monitor the read end of each socket pair, declare your Unix signal handlers to be static class methods, and declare a slot function corresponding to each of your Unix signal handlers. In this example, we intend to handle both the SIGHUP and SIGTERM signals. Note: You should read the socketpair(2) and the sigaction(2) man pages before plowing through the following code snippets.

tresf commented:
> QSocketNotifier

I read that article and I can't for the life of me understand why a network class is being used to send an interrupt to the main app.

Quoting that article:

> You should read the `socketpair(2)` and the `sigaction(2)` man pages before plowing through the following code snippets.

Is the intent that Qt wants to be able to handle signals that do more than simply interrupt the application?  Or signals that aren't covered in this tutorial?  Why would a developer put so many layers of obscurity between the terminal and the calling of exit?  What problem does this  solve?

tresf commented:
> why a network class is being used to send an interrupt to the main app.

Here's a write-up on it:  https://stackoverflow.com/a/34467910/3196753

Apparently Qt's solution relies on [`::write(...)` specifically being a signal-safe function](https://man7.org/linux/man-pages/man7/signal-safety.7.html).   I'll throw something together that follows this design.

tresf commented:
The new implementation is done, but it's completely overshadowed by `Engine.cpp` and the scanning of back-ends.  If there's a way to put precedence please share.

tresf commented on src/gui/GuiApplication.cpp:
```diff
@@ -73,8 +79,11 @@ GuiApplication* GuiApplication::instance()
 
 
-GuiApplication::GuiApplication()
+GuiApplication::GuiApplication(int* sigintFd)
```
If there's a better way to pass this in without modifying the constructor, please let me know.   I expected static functions to handle this, but they behave different in C++ versus other languages.  I tried `extern`, but that wanted a member variable.  I'd much rather not have a separate constructor for Windows.

tresf commented:
I've re-requested some reviews.  If I don't receive reviews, I'll merget this as-is in a copuld weeks.

rubiefawn reviewed (commented):
I've adjusted it so that the file descriptors are stored in a static member rather than being passed into the constructor.

tresf commented:
> I've adjusted it so that the file descriptors are stored in a static member rather than being passed into the constructor.

Thank you!  I took it a step further and just moved all of the handling in GuiApplication. 🤞

tresf commented:
Some slight refactoring, I think this PR is ready for review now.

tresf commented:
@rubiefawn @messmerd @sakertooth if I can get a sign-off before merging it would be greatly appreciated.  If not, I'll likely merge this without approval in about a week.
